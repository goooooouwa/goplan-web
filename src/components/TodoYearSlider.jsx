import { Grid } from '@mui/material';
import httpService from 'httpService';
import { calculatedEndDate } from 'lib/timeLeft';
import moment from 'moment';
import React, { useState } from 'react';
import SliderContainer from './SliderContainer';
import TodoItem from './TodoItem';

const rangeMin = 0;
const rangeMax = 11;

export default function TodoYearSlider(props) {
  const [startDate, setStartDate] = useState((props.todo.startDate !== null) ? moment(props.todo.startDate) : moment());
  const [endDate, setEndDate] = useState((props.todo.endDate !== null) ? moment(props.todo.endDate) : calculatedEndDate(props.todo.startDate, props.todo.timeSpan));
  const selectedYear = moment(props.selectedYear);

  const isInRange = (date) => {
    return date.isValid() && (date.isAfter(selectedYear.startOf("year")) && date.isBefore(selectedYear.endOf("year")));
  };

  const rangeMark = (date) => {
    if (date.isValid()) {
      let rangeMark;
      if (date.isBefore(selectedYear.startOf("year"))) {
        rangeMark = rangeMin;
      } else if (date.isAfter(selectedYear.endOf("year"))) {
        rangeMark = rangeMax;
      } else {
        rangeMark = date.month();
      }
      return rangeMark;
    } else {
      return -1;
    }
  };

  function handleMonthChange(months) {
    const [newStartMonth, newEndMonth] = months;
    const todoData = {
      project_id: props.todo.projectId,
      name: props.todo.name,
      description: props.todo.description,
      time_span: props.todo.timeSpan,
      start_date: startDate.month(newStartMonth).toISOString(),
      end_date: endDate.month(newEndMonth).toISOString(),
      repeat: props.todo.repeat,
      repeat_period: props.todo.repeatPeriod,
      repeat_times: props.todo.repeatTimes,
      instance_time_span: props.todo.instanceTimeSpan
    };

    httpService.put(`/todos/${props.todo.id}.json`, todoData)
      .then((response) => {
        setStartDate((response.data.startDate !== null) ? moment(response.data.startDate) : moment());
        setEndDate((response.data.endDate !== null) ? moment(response.data.endDate) : calculatedEndDate(response.data.startDate, response.data.timeSpan));
      })
      .catch(function (error) {
        console.log(error);
      });
  }

  return (
    <>
      <Grid item xs={12} md={4}>
        <TodoItem todo={props.todo} />
      </Grid>
      <Grid item xs={12} md={8} sx={{ px: 3 }}>
        <SliderContainer
          marks={props.marks}
          rangeMin={rangeMin}
          rangeMax={rangeMax}
          rangeStart={rangeMark(startDate)}
          rangeEnd={rangeMark(endDate)}
          disableRangeStart={!isInRange(startDate)}
          disableRangeEnd={!isInRange(endDate)}
          handleChange={handleMonthChange}
        />
      </Grid>
    </>
  );
}
