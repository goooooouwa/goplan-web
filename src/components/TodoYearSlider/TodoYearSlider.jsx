import { Grid } from '@mui/material';
import isInYearRange from 'utils/rangeCheck';
import moment from 'moment';
import React from 'react';
import PropTypes from 'prop-types';
import momentPropTypes from 'react-moment-proptypes';
import SHARED_PROP_TYPES from 'utils/sharedPropTypes';
import TimelineRangeSlider from 'components/TimelineRangeSlider/TimelineRangeSlider';
import TodoItem from 'components/TodoItem/TodoItem';
import TimelineSlider from 'components/TimelineSlider/TimelineSlider';

const rangeMin = 0;
const rangeMax = 11;

export default function TodoYearSlider(props) {
  const startDate = (props.todo.startDate !== null) ? moment(props.todo.startDate) : moment();
  const endDate = (props.todo.endDate !== null) ? moment(props.todo.endDate) : moment();

  const rangeMark = (date) => {
    if (date.isValid()) {
      let rangeMark;
      if (date.isBefore(props.selectedYear.clone().startOf("year"))) {
        rangeMark = rangeMin;
      } else if (date.isAfter(props.selectedYear.clone().endOf("year"))) {
        rangeMark = rangeMax;
      } else {
        rangeMark = date.month();
      }
      return rangeMark;
    } else {
      return -1;
    }
  };

  return (
    <>
      <Grid item xs={12} md={4}>
        <TodoItem todo={props.todo} handleTodoChange={props.handleTodoChange}/>
      </Grid>
      <Grid item xs={12} md={8} sx={{ px: 3 }}>
        {props.todo.repeat &&
          <TimelineRangeSlider
            marks={props.marks}
            rangeMin={rangeMin}
            rangeMax={rangeMax}
            rangeStart={rangeMark(startDate)}
            rangeEnd={rangeMark(endDate)}
            disableRangeStart={!isInYearRange(startDate, props.selectedYear)}
            disableRangeEnd={!isInYearRange(endDate, props.selectedYear)}
            handleChangeCommited={(newValue) => { props.handleMonthChange(props.todo, newValue) }}
          />
        }
        {!props.todo.repeat &&
          <TimelineSlider
            marks={props.marks}
            rangeMin={rangeMin}
            rangeMax={rangeMax}
            rangeStart={rangeMark(startDate)}
            disableRangeStart={!isInYearRange(startDate, props.selectedYear)}
            handleChangeCommited={(newValue) => { props.handleMonthChange(props.todo, newValue) }}
          />
        }
      </Grid>
    </>
  );
}

TodoYearSlider.propTypes = {
  selectedYear: momentPropTypes.momentObj.isRequired,
  marks: SHARED_PROP_TYPES.marks,
  todo: SHARED_PROP_TYPES.todo,
  handleTodoChange: PropTypes.func.isRequired,
  handleMonthChange: PropTypes.func.isRequired
};