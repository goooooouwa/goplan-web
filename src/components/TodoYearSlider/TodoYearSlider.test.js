import moment from 'moment';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import TodoYearSlider from '../TodoYearSlider/TodoYearSlider';

const marks = [
  {
    value: 0,
    label: 'Jan',
  },
  {
    value: 1,
    label: 'Feb',
  },
  {
    value: 2,
    label: 'Mar',
  },
  {
    value: 3,
    label: 'Apr',
  },
  {
    value: 4,
    label: 'May',
  },
  {
    value: 5,
    label: 'Jun',
  },
  {
    value: 6,
    label: 'Jul',
  },
  {
    value: 7,
    label: 'Aug',
  },
  {
    value: 8,
    label: 'Sep',
  },
  {
    value: 9,
    label: 'Oct',
  },
  {
    value: 10,
    label: 'Nov',
  },
  {
    value: 11,
    label: 'Dec',
  },
];


const todo = {
  id: 1,
  projectId: 1,
  status: false,
  repeat: false,
  name: "Todo 1",
  startDate: '2022-01-02',
  endDate: '2022-01-02',
  instanceTimeSpan: 1,
};

it('should show a timeline slider', () => {
  const component = renderer.create(
    <BrowserRouter>
      <TodoYearSlider todo={todo} marks={marks} selectedYear={moment('2022-01-02')} handleTodoChange={() => { }} handleMonthChange={() => { }} />
    </BrowserRouter>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});