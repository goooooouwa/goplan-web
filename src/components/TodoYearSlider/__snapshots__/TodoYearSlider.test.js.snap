// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show a timeline slider 1`] = `
Array [
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 css-13i4rnv-MuiGrid-root"
  >
    <li
      className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
      disabled={false}
    >
      <span
        className="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <input
          aria-label="controlled"
          checked={false}
          className="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate={false}
          onChange={[Function]}
          type="checkbox"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </span>
      <a
        className="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        href="/projects/1/todos/1"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
      >
        <div
          className="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Todo 1
          </span>
        </div>
      </a>
    </li>
  </div>,
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8 css-5swanh-MuiGrid-root"
  >
    <span
      className="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root css-101tn95-MuiSlider-root"
      onMouseDown={[Function]}
    >
      <span
        className="MuiSlider-rail css-14pt78w-MuiSlider-rail"
      />
      <span
        className="MuiSlider-track css-1gv0vcd-MuiSlider-track"
        style={
          Object {
            "left": "0%",
            "width": "0%",
          }
        }
      />
      <span
        className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
        data-index={0}
        style={
          Object {
            "left": "0%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={1}
        style={
          Object {
            "left": "9.090909090909092%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={2}
        style={
          Object {
            "left": "18.181818181818183%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={3}
        style={
          Object {
            "left": "27.272727272727273%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={4}
        style={
          Object {
            "left": "36.36363636363637%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={5}
        style={
          Object {
            "left": "45.45454545454545%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={6}
        style={
          Object {
            "left": "54.54545454545455%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={7}
        style={
          Object {
            "left": "63.63636363636363%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={8}
        style={
          Object {
            "left": "72.72727272727273%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={9}
        style={
          Object {
            "left": "81.81818181818181%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={10}
        style={
          Object {
            "left": "90.9090909090909%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={11}
        style={
          Object {
            "left": "100%",
          }
        }
      />
      <span
        className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
        data-index={0}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "left": "0%",
            "pointerEvents": undefined,
          }
        }
      >
        <input
          aria-orientation="horizontal"
          aria-valuemax={11}
          aria-valuemin={0}
          aria-valuenow={0}
          aria-valuetext={0}
          data-index={0}
          disabled={false}
          max={11}
          min={0}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          step={1}
          style={
            Object {
              "border": 0,
              "clip": "rect(0 0 0 0)",
              "direction": "ltr",
              "height": "100%",
              "margin": -1,
              "overflow": "hidden",
              "padding": 0,
              "position": "absolute",
              "whiteSpace": "nowrap",
              "width": "100%",
            }
          }
          type="range"
          value={0}
        />
        <span
          aria-hidden={true}
          className="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
        >
          <span
            className="MuiSlider-valueLabelCircle"
          >
            <span
              className="MuiSlider-valueLabelLabel"
            >
              Jan
            </span>
          </span>
        </span>
      </span>
    </span>
  </div>,
]
`;
